src/main/java/bio/terra/service/filedata/FileMetadataUtils.java

```diff
@@ -77,10 +80,33 @@ public FireStoreDirectoryEntry makeDirectoryEntry(String lookupDirPath) {
   public String makeLookupPath(String fullPath) {
     String temp = StringUtils.prependIfMissing(fullPath, "/");
     temp = StringUtils.removeEnd(temp, "/");
-    return ROOT_DIR_NAME + temp;
+    temp = StringUtils.prependIfMissing(temp, ROOT_DIR_NAME);
+    return temp;
   }
 
   public String makePathFromLookupPath(String lookupPath) {
     return StringUtils.removeStart(lookupPath, ROOT_DIR_NAME);
   }
+
+  public List<String> findNewDirectoryPaths(
+      List<FireStoreDirectoryEntry> datasetEntries, LRUMap<String, Boolean> pathMap) {
+
+    List<String> pathsToCheck = new ArrayList<>();
+    for (FireStoreDirectoryEntry entry : datasetEntries) {
+      // Only probe the real directories - not leaf file reference or the root
+      String lookupDirPath = makeLookupPath(entry.getPath());
+      for (String testPath = lookupDirPath;
+          !testPath.isEmpty() && !testPath.equals(FileMetadataUtils.ROOT_DIR_NAME);
+          testPath = getDirectoryPath(testPath)) {
+
+        // check the cache
+        if (pathMap.get(testPath) == null) {
+          // not in the cache: add to checklist and a to cache
+          pathsToCheck.add(testPath);
+          pathMap.put(testPath, true);
+        }
```

Comment:
I think a better way to express this is to use `computeIfAbsent()`, which has the additional feature of being thread safe.
```suggestion
        pathMap.computeIfAbsent(testPath, s -> {
          pathsToCheck.add(s);
          return true;
        });
```

Reply:
Same applies here - https://github.com/DataBiosphere/jade-data-repo/pull/1073/files#r720421651